Handle node info updates

marcelveldt opened the pull request:
When node info (such as model/name) updates, push that info to the device registry. Usecase in particular is adding a new Z-Wave node to the mesh where the detailed nodeinfo is lazy loaded.

Fix for #68

marcelveldt committed:
handle node info updates

When node info (such as model/name) updates, push that info to the device registry. Usecase in particular is adding a new Z-Wave node to the mesh where the detailed nodeinfo is lazy loaded.

balloob commented on custom_components/zwave_mqtt/__init__.py:
```diff
@@ -246,6 +250,35 @@ async def handle_remove_node(hass: HomeAssistant, node_id: int):
+    dev_registry = await get_dev_reg(hass)
+    # grab device in device registry attached to this node
+    device = dev_registry.async_get_device([(DOMAIN, node.id)], [])
+    if device:
```
Let's use a guard clause.

```python
if not device:
    return
```

balloob commented on custom_components/zwave_mqtt/__init__.py:
```diff
@@ -246,6 +250,35 @@ async def handle_remove_node(hass: HomeAssistant, node_id: int):
+        # update device in device registry with (updated) info
+        # also handle any slave devices (node instances)
+        for item in dev_registry.devices.values():
+            if item.id == device.id or item.via_device_id == device.id:
```
Use a guard clause here too

```python
if item.id != device.id and item.via_device_id != device.id:
    continue
```

balloob commented on custom_components/zwave_mqtt/__init__.py:
```diff
@@ -246,6 +250,35 @@ async def handle_remove_node(hass: HomeAssistant, node_id: int):
+            if item.id == device.id or item.via_device_id == device.id:
+                dev_name = f"{node.node_manufacturer_name} {node.node_product_name}"
+                if item.via_device_id:
+                    instance = list(item.identifiers)[0][2]
```
You cannot expect this to be in any order. You need to lookup based on the key.

marcelveldt replied:
You're right, I took the shortcut. Will adjust

balloob commented on custom_components/zwave_mqtt/__init__.py:
```diff
@@ -246,6 +250,35 @@ async def handle_remove_node(hass: HomeAssistant, node_id: int):
+        # also handle any slave devices (node instances)
+        for item in dev_registry.devices.values():
+            if item.id == device.id or item.via_device_id == device.id:
+                dev_name = f"{node.node_manufacturer_name} {node.node_product_name}"
```
Home Assistant already stores manufacturer and product. Seems weird to put in device name too?

marcelveldt replied:
This is used as the default name the device gets but you're right, this does not have to be the full devicename and manufacturer. We can also choose some shorter default name like "Z-Wave Node 1"

marcelveldt replied:
I'm working on another PR that has focus on device/entity names to make things more user friendly. I will take this naming improvement there. So leave it here for now to narrow down scope.

balloob commented on custom_components/zwave_mqtt/__init__.py:
```diff
@@ -246,6 +250,35 @@ async def handle_remove_node(hass: HomeAssistant, node_id: int):
+        for item in dev_registry.devices.values():
+            if item.id == device.id or item.via_device_id == device.id:
+                dev_name = f"{node.node_manufacturer_name} {node.node_product_name}"
+                if item.via_device_id:
```
Isn't this always the case, because all Z-Wave devices are connected via the OZW Daemon ?

marcelveldt replied:
@balloob no, this is used for the Instances stuff we talked about. Nodes with multiple instances get split up into multiple virtual devices.

balloob commented on custom_components/zwave_mqtt/__init__.py:
```diff
@@ -246,6 +250,35 @@ async def handle_remove_node(hass: HomeAssistant, node_id: int):
+                    instance = list(item.identifiers)[0][2]
+                    dev_name += f" - Instance {instance}"
+                # TEMP: Using private update method as public uodate needs extending.
+                # TODO: Use public function as soon as core is adjusted,
```
This has been added https://github.com/home-assistant/core/pull/34500

marcelveldt replied:
Nice, that's quick :-)

marcelveldt replied:
What to do ? Adjust code and not merge until a new hass version comes out ?
Or merge like it is now and insert a separate PR at the same time to adjust the private to pubic method ?

As we're publishing this into HACS we need to make sure we do not publish code depending om core changes

MartinHjelmare replied:
How does the HACS publish work. Does it use current master branch or latest github release?

marcelveldt replied:
latest GH release. I'd like to do a new release today or tomorrow with yesterday's fixes (and the PR I'm about to do)

cgarwood replied:
In hacs.json in the root of the repo you can add a minimum home assistant version: `"homeassistant": "0.109.0",`

MartinHjelmare replied:
What happens if that home assistant version isn't released yet? Does the HACS integration not install then? Can you select different HACS integration versions?

cgarwood replied:
Not sure how it will handle an HA version that hasn't been released yet. @ludeeus ?

ludeeus replied:
It will tell the user to upgrade

MartinHjelmare replied:
Should we require 0.109.0, then merge this and wait to release another version here until 0.109.0 is released on Wednesday?

marcelveldt replied:
Does 0.109 contain the fix ? In that case we'll wait with merging, that's only a few days from now.

MartinHjelmare replied:
Yeah, the fix is in rc branch.

marcelveldt replied:
Code adjusted to use the public function (released in Hass 0.109). We should only make sure that 0.109 is set as minimum version for next release on HACS.

marcelveldt committed:
cleanup

a few adjustments discussed in review

marcelveldt committed:
simplify

the instance stuff is not needed at this stage, only makes it really complex at this stage

MartinHjelmare commented on custom_components/zwave_mqtt/__init__.py:
```diff
@@ -90,6 +91,9 @@ def async_node_added(node):
         _LOGGER.debug("[NODE CHANGED] node_id: %s", node.id)
         data_nodes[node.id] = node
+        # notify devices about the node change
+        if node.id not in removed_nodes:
```
A bit off topic: When is `removed_nodes` cleared or node_ids removed from `removed_nodes`? It seems we only append to this list but never remove ids.

Should we remove the node id from this list after we have removed the entities of the node? If the user adds the same node back again the node id should no longer be in the removed list, I think.

marcelveldt replied:
Yeah, but you will never ever get the same node id when you add a new node. But for completeness I'll add it in ;-)

MartinHjelmare replied:
Oh. So node ids are not stable for the same device?

MartinHjelmare replied:
Node id is just a position in the mesh?

kpine replied:
The controller will not re-assign a node ID unless you exceed the maximum (232) and then it will wrap back around and re-use any removed ones.

Fishwaldo commented:
Unless you use replacefailednode. Then it will reuse the Node ID on the *failed* node

marcelveldt commented:
@MartinHjelmare and @cgarwood are you guys OK with merging this one because it improves the usability ? I will submit a new PR to change the private to the pubic method. Otherwise we're blocked until a new hass release comes out with this change applied.

cgarwood commented:
I think its fine to merge as-is and then update it once the new HA release comes out

marcelveldt committed:
switch to public method for updating device info

marcelveldt committed:
require homeassistant 0.109.0 as minimal version

We use a feature to update the device registry added in hass 0.109.0
